File: Python/lld-examples/medium/online-shopping-system/product_catalog.py
from typing import List, Dict, Optional
from product import Product

class ProductCatalog:
    """Product catalog class for managing products"""
    
    def __init__(self):
        self.products: Dict[str, Product] = {}
        self.products_by_name: Dict[str, Product] = {}
    
    def add_product(self, product: Product) -> bool:
        """Add product to catalog"""
        if product.product_id in self.products:
            return False  # Product already exists
        
        self.products[product.product_id] = product
        self.products_by_name[product.name.lower()] = product
        return True
    
    def get_product_by_id(self, product_id: str) -> Optional[Product]:
        """Get product by product ID"""
        return self.products.get(product_id)
    
    def get_product_by_name(self, name: str) -> Optional[Product]:
        """Get product by name"""
        return self.products_by_name.get(name.lower())
    
    def get_all_products(self) -> List[Product]:
        """Get all products"""
        return list(self.products.values())
    
    def search_products(self, keyword: str) -> List[Product]:
        """Search products by keyword"""
        keyword = keyword.lower()
        results = []
        
        for product in self.products.values():
            if (keyword in product.name.lower() or 
                keyword in product.description.lower()):
                results.append(product)
        
        return results
    
    def get_products_by_category(self, category: str) -> List[Product]:
        """Get products by category (simplified - could be enhanced)"""
        # This is a simplified implementation
        return [p for p in self.products.values() if category.lower() in p.name.lower()]
    
    def update_product_stock(self, product_id: str, stock: int) -> bool:
        """Update product stock"""
        product = self.get_product_by_id(product_id)
        if product:
            product.stock = stock
            return True
        return False
    
    def update_product_price(self, product_id: str, price: float) -> bool:
        """Update product price"""
        product = self.get_product_by_id(product_id)
        if product:
            product.price = price
            return True
        return False
    
    def remove_product(self, product_id: str) -> bool:
        """Remove product from catalog"""
        product = self.get_product_by_id(product_id)
        if product:
            del self.products[product_id]
            del self.products_by_name[product.name.lower()]
            return True
        return False
    
    def get_available_products(self) -> List[Product]:
        """Get available products only"""
        return [p for p in self.products.values() if p.is_available and p.stock > 0]
    
    def get_total_products(self) -> int:
        """Get total number of products"""
        return len(self.products) 

File: JAVA/lld-examples/medium/online-shopping-system/src/main/java/com/systemdesign/lld/shopping/ProductCatalog.java
package com.systemdesign.lld.shopping;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * ProductCatalog manages the product inventory and catalog operations
 */
public class ProductCatalog {
    private Map<String, Product> products;
    private Map<String, List<Product>> categoryProducts;
    
    public ProductCatalog() {
        this.products = new ConcurrentHashMap<>();
        this.categoryProducts = new ConcurrentHashMap<>();
    }
    
    /**
     * Add a new product to the catalog
     */
    public boolean addProduct(Product product) {
        if (products.containsKey(product.getProductId())) {
            return false; // Product already exists
        }
        
        products.put(product.getProductId(), product);
        
        // Add to category mapping
        String category = product.getCategory();
        categoryProducts.computeIfAbsent(category, k -> new ArrayList<>()).add(product);
        
        return true;
    }
    
    /**
     * Remove a product from the catalog
     */
    public boolean removeProduct(String productId) {
        Product product = products.remove(productId);
        if (product != null) {
            // Remove from category mapping
            String category = product.getCategory();
            List<Product> categoryList = categoryProducts.get(category);
            if (categoryList != null) {
                categoryList.remove(product);
                if (categoryList.isEmpty()) {
                    categoryProducts.remove(category);
                }
            }
            return true;
        }
        return false;
    }
    
    /**
     * Get product by ID
     */
    public Product getProduct(String productId) {
        return products.get(productId);
    }
    
    /**
     * Get all products
     */
    public List<Product> getAllProducts() {
        return new ArrayList<>(products.values());
    }
    
    /**
     * Get products by category
     */
    public List<Product> getProductsByCategory(String category) {
        return categoryProducts.getOrDefault(category, new ArrayList<>());
    }
    
    /**
     * Search products by name
     */
    public List<Product> searchProductsByName(String searchTerm) {
        return products.values().stream()
                .filter(product -> product.getName().toLowerCase().contains(searchTerm.toLowerCase()))
                .collect(Collectors.toList());
    }
    
    /**
     * Get products within price range
     */
    public List<Product> getProductsByPriceRange(double minPrice, double maxPrice) {
        return products.values().stream()
                .filter(product -> product.getPrice() >= minPrice && product.getPrice() <= maxPrice)
                .collect(Collectors.toList());
    }
    
    /**
     * Get all available categories
     */
    public Set<String> getAllCategories() {
        return new HashSet<>(categoryProducts.keySet());
    }
    
    /**
     * Update product stock
     */
    public boolean updateProductStock(String productId, int newStock) {
        Product product = products.get(productId);
        if (product != null) {
            product.updateStock(newStock);
            return true;
        }
        return false;
    }
    
    /**
     * Check if product is available
     */
    public boolean isProductAvailable(String productId, int quantity) {
        Product product = products.get(productId);
        return product != null && product.isAvailable(quantity);
    }
    
    /**
     * Get low stock products (stock < threshold)
     */
    public List<Product> getLowStockProducts(int threshold) {
        return products.values().stream()
                .filter(product -> product.getStock() < threshold)
                .collect(Collectors.toList());
    }
    
    /**
     * Get total number of products
     */
    public int getTotalProducts() {
        return products.size();
    }
    
    /**
     * Get catalog statistics
     */
    public Map<String, Object> getCatalogStats() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("totalProducts", products.size());
        stats.put("totalCategories", categoryProducts.size());
        stats.put("lowStockProducts", getLowStockProducts(5).size());
        stats.put("totalValue", products.values().stream()
                .mapToDouble(p -> p.getPrice() * p.getStock())
                .sum());
        
        return stats;
    }
} 